client_code/kompot/_persist.py

```diff
@@ -0,0 +1,279 @@
+            try:
+                return getattr(self._dispatcher, name)
+            except AttributeError:
+                return self._row[name]
+        else:
+            return self._row[name]
+
+    def __setattr__(self, name, value):
+        if name.startswith("_") or not self._initialised:
+            object.__setattr__(self, name, value)
+        else:
+            setattr(self._dispatcher, name, value)
+
+    def get(self, *args, **kwargs):
+        "Fetch a data tables row and make it available to the parent object"
+        self._row = self._server_functions["getter"](*args, **kwargs)
+        self._dispatcher.clear()
+
+    def create(self):
+        """Create a new data tables row from the parent object"""
+        result = self._server_functions["creator"](changes=serialize(self._dispatcher))
+        self._dispatcher.clear()
+        return result
+
+    def update(self):
+        """Save changes made to the parent object to the data tables row"""
+        result = self._server_functions["updater"](
+            changes=serialize(self._dispatcher), row=self._row
+        )
+        self._dispatcher.clear()
+        return result
+
+    def delete(self):
+        """Delete the data tables row"""
+        result = self._server_functions["deleter"](row=self._row)
+        self._dispatcher.clear()
+        return result
+
+    def reset(self):
+        self._dispatcher.clear()
+
+
+def _get_value(self, key):
+    if not key.startswith("_"):
+        try:
+            return getattr(self._store, key)
+        except AttributeError:
+            raise AttributeError(
+                f"{self.__class__.__name__} instance has no store defined"
+            )
+
+
+def _set_value(self, key, value):
+    is_private = key.startswith("_")
+    is_linked_attribute = hasattr(self.__class__, key) and isinstance(
+        getattr(self.__class__, key), LinkedAttribute
+    )
+    if is_private or is_linked_attribute:
+        object.__setattr__(self, key, value)
+    else:
+        try:
+            setattr(self._store, key, value)
+        except AttributeError:
+            raise AttributeError(
+                f"{self.__class__.__name__} instance has no store defined"
+            )
+
+
+def _get(self, *args, **kwargs):
+    return self._store.get(*args, **kwargs)
+
+
+def _create(self):
+    return self._store.create()
+
+
+def _save(self):
+    return self._store.save()
+
+
+def _delete(self):
+    return self._store.delete()
+
+
+def _reset(self):
+    self._store.reset()
+
+
+MEMBERS = {
+    "__getattr__": _get_value,
+    "__getitem__": _get_value,
+    "__setattr__": _set_value,
+    "__setitem__": _set_value,
+    "get": _get,
+    "create": _create,
+    "save": _save,
+    "delete": _delete,
+    "reset": _reset,
+}
+
+
+def persisted_class(cls):
+    """A decorator for a class with a persistence mechanism"""
+    return type(cls.__name__, (object,), dict(cls.__dict__, **MEMBERS))
+
+
+@property
+def _server_functions(self):
```

Comment:
I used a property because I have a (marginal) preference for `self._server_functions["getter"]` rather than `self._server_functions()["getter"]`

Reply:
sorry I meant the following

```
def row_backed_class(cls):
    """A decorator for a class persisted by a data tables row"""
    _cls = persisted_class(cls)
    setattr(_cls, "_server_functions", combine_server_functions())
    setattr(_cls, "__init__", _row_backed_init)
    return _cls
```
Where `combine_server_functions()` is the current implementation of `_server_functions`

Reply:
I'm less keen on the property because, although we only access `_server_functions` once in this implementation, it's quite a lot of work if we were to regularly access `self._server_functions`

And actually we do the work on each instance creation
When we can probably do the work once on cls creation

Reply:
yeah, that's much nicer. I'll fix that up...